### PR TITLE
Remove mappings in page cache lazily

### PR DESCRIPTION
**Pull request by arpad-m**
## Problem

Right now, there are mapping removals for the page cache in the `Drop` destructors in two places:

* `PageWriteGuard` for pages with `valid == false` that haven't been written to yet
* `EphemeralFile` for all pages associated with the file being dropped

Both these mapping removals lock slots in the page cache and then free up page slots related to the datastructures that are being dropped. In #5023, we want to make these slot locks async, and as we can't make the `Drop` destructors async themselves, we need to find alternatives.

One possibility would be just not do any removals, which would still be perfectly correct, but the pages would now be "leaked" and eviction code would treat them the same as currently used pages. This would obviously be quite unfortunate.

That's why we now instead of eagerly removing the mappings, we mark the pages for removal. Later we remove the mappings, once we are in the need for new pages in the `find_victim` function.

## Summary of changes

There are a few smaller refactors:

* use `Vec::leak` instead of `Vec::into_boxed_slice` plus `Box::leak`
* introduce a `set_usage_count` function instead of manually calling `usage_count.store`
* improve comments for the `valid` field, they were quite terse before. 

The main changes are:

* introduce a `slot_idx` field of `SlotInner` which is used to access the corresponding `Slot` that has the `usage_count` field (direct pointers to the parent are hard in Rust).
* reset the `usage_count` field in the `PageWriteGuard` destructor. The `find_victim` function will then treat this page as having zero users (as it should) and remove its mappings.
* as we don't know the particular slot indices associated with the ephemeral file, we use a queue to remove their mappings from the `find_victim` function.

See also discussion in https://github.com/neondatabase/neon/pull/5023#discussion_r1304194495 which outlines the alternatives and design constraints, especially why we chose locks from `crossbeam_channel` instead of locks from std or tokio.

Note that metrics wise, the delayed execution of the `remove_mappings` function is not without effect: as the `remove_mappings` function updates the metrics, and it is executed now in a delayed fashion, metrics will show some pages as used even though they are basically unused and will be freed at the next time there is need for new pages.

Part of #4743.

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: pageserver/src/page_cache.rs [hunk 1]
```diff
@@ -773,6 +790,7 @@ impl PageCache {
     ///
     /// On return, the slot is empty and write-locked.
     fn find_victim(&self) -> anyhow::Result<(usize, RwLockWriteGuard<SlotInner>)> {
+        self.empty_queued_file_evictions();
```
Comment (jcsp) [inline pageserver/src/page_cache.rs:793]: The try_recv in crossbeam does some atomics -- we should avoid calling it on every find_victim.  Something very simple might work here, like only calling it once we have seen at least one `WouldBlock`.

File: pageserver/src/page_cache.rs [hunk 2]
```diff
@@ -299,17 +310,16 @@ impl PageWriteGuard<'_> {
...
+    /// from the page cache.
     fn drop(&mut self) {
         assert!(self.inner.key.is_some());
         if !self.valid {
-            let self_key = self.inner.key.as_ref().unwrap();
-            PAGE_CACHE.get().unwrap().remove_mapping(self_key);
-            self.inner.key = None;
+            // Make sure that the page gets freed at the next time we look for a
+            // victim page.
+            PAGE_CACHE.get().unwrap().slots[self.inner.slot_idx].set_usage_count(0);
```
Comment (jcsp) [inline pageserver/src/page_cache.rs:322]: Nice.

File: Cargo.lock [hunk 3]
```diff
@@ -2615,6 +2615,7 @@ dependencies = [
  "consumption_metrics",
  "crc32c",
  "criterion",
+ "crossbeam-channel",
```
Comment (jcsp) [inline Cargo.lock:2618]: I don't have a problem with including crossbeam (it is a very reputable crate), but I think we could manage with a tokio channel if we wanted to.

The receiver try_recv requires &mut, but that can be addressed by spawning a task that owns the receiver, and has a reference to the page cache.  Then that task just sits in a receive loop, and calls into page cache to process the entries with `drop_buffers_for_immutable`

File: pageserver/src/page_cache.rs (hunk 1 above)
Comment (jcsp) [reply pageserver/src/page_cache.rs:793]: (or if we use a background task to drain the queue, this problem goes away)

Comment (arpad-m) [reply pageserver/src/page_cache.rs:793]: WDYM by having seen at least one `WouldBlock`? The same `start_recv` function that is used by `try_recv` is also used by the blocking `recv` function.

Comment (arpad-m) [reply pageserver/src/page_cache.rs:793]: Also, not really a fan of having a background job do this, because of the extra complexity. Wondering what others think about having a background job.

Comment (jcsp) [reply pageserver/src/page_cache.rs:793]: > WDYM by having seen at least one WouldBlock? The same start_recv function that is used by try_recv is also used by the blocking recv function.

I mean giving the loop later in the function a chance to find a slot, and only going into the queue-draining path if we find that we aren't immediately finding an available slot.

One could also have some heuristic like not draining the queue until we've scanned 10 slots, or just draining the queue on every Nth call into the function.

Basically I'm looking at find_victim as a hot path, where we should avoid adding calls into synchronization primitives like the queue receiver.

Comment (arpad-m) [reply pageserver/src/page_cache.rs:793]: There is a tradeoff here, the longer we wait with first poking at existing slots, the more the clock algorithm reduces usage counts across the slots. In other words, slots are moved towards freeing that wouldn't be needed to be freed.

Also note that the `find_victim` function already does access atomics internally for each candidate slot.

**Comment by problame** [comment]
My PR #5119 eliminates 

> PageWriteGuard for pages with valid == false that haven't been written to yet

Now looking into the EphemeralFile case.

**Comment by arpad-m** [comment]
closing in favour of #5122 (we can always reopen if that approach doesn't work).
